rfcs/0005-nix-encryption.md

```diff
@@ -0,0 +1,273 @@
+
+# Motivation
+[motivation]: #motivation
+
+It is often necessary to store secrets such as passwords or SSL
+certificates in NixOS system closures. Unfortunately, the Nix store is
+world-readable, so this is currently not possible in a secure way;
+anybody who gains access to the filesystem of a NixOS machine can read
+any file in its Nix store. Nevertheless, lots of NixOS modules *do*
+store secrets in the Nix store (see
+https://github.com/NixOS/nixpkgs/issues/24288), many of them lacking
+warnings about the danger of doing so.
+
+# Detailed design
+[design]: #detailed-design
+
+The proposal consists of the following components:
+
+* A builtin function `encryptString :: Path -> String -> String` that
+  encrypts a string using a symmetric key stored in the specified
+  path. The encryption is done using libsodium's authenticated
+  encryption function `crypto_secretbox_easy()`. The resulting string
+  has the format `<{|nixcrypt:base64-encoded-data|}>`, where
+  `base64-encoded-data` is the base-64-encoded output of
+  `crypto_secretbox_easy()`, along with the nonce used.
+
+* A command `nix decrypt` to decrypt files containing encrypted
+  strings produced by `encryptString`. This command searches for
+  `<{|nixcrypt:...|}>` fragments and decrypts them using a decryption
+  key specified on the command line. Any other data is copied
+  verbatim.
+
+* A command `nix generate-key` to generate a key in the format
+  expected by `encryptString` and `nix decrypt`.
+
+A typical usage in a module NixOS would look like this:
+
+```
+let
+  configFile = pkgs.writeText "foo.conf"
+    ''
+      # Store the password of the foo service in encrypted form.
+      password=${builtins.encryptString <nixos-store-key> cfg.password}
+    '';
+in {
+
+  systemd.services.foo = {
+    preStart =
+      ''
+        # Decrypt the configuration file.
+        ${config.nix.package}/bin/nix-store \
+          --decrypt ${toString <nixos-store-key>} \
+          ${configFile} > /run/secret/foo.conf
+      '';
+    # Run the service using the decrypted configuration file/
+    serviceConfig.Exec =
+      "${pkgs.foo}/bin/foo -c /run/secret/foo.conf";
+  };
+};
+```
+(See below for a discussion of simplifications to this example.)
+
+At evaluation time, this causes the following string to be emitted by
+the Nix evaluator:
+```
+password=<{|nixcrypt:base64-encoded-data|}>
+```
+
+Note that we don't necessarily encrypt entire files (though that's
+possible); generally, we'll encrypt only the sensitive parts of
+files. This allows users to read non-sensitive parts of configuration
+files. It *may* also enable graceful degradation if secrets cannot be
+decrypted, but this is risky (see below under "Decryption failure").
+
+The assumption is that `<nixos-store-key>` resolves to something like
+`/var/lib/nixos/store-key`, which should contain a key generated by
+`nix generate-key` and should obviously be readable by root only.
```

Comment:
@zimbatm: I believe this is safe in the presence of both binary caches and distributed builds.

For binary caches, the actual cached configuration only stores encrypted secrets, so even if you mirror the configuration to other `/nix/store`s they won't be able to decrypt these secrets unless they have access the same key.

For distributed builds, files on the `NIX_PATH` are not copied to distributed build slaves, so if the `<nixos-store-key>` path is missing on the build slave machine(s) then the build will just fail.